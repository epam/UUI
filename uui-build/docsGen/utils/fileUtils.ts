import fs from 'fs';
import path from 'path';
import { uuiRoot } from '../constants';

export function makeRelativeToUuiRoot(fullPath: string) {
    return removeTrailingDots(path.relative(uuiRoot, fullPath).replace(/\\/g, '/'));
}

export function saveContentToFile(fullPath: string, contentToSave: object | string) {
    if (fs.existsSync(fullPath)) {
        fs.rmSync(fullPath, { force: true });
    }
    const content = typeof contentToSave === 'string' ? contentToSave : JSON.stringify(contentToSave, undefined, 1);
    fs.writeFileSync(fullPath, content);
}

export function isExternalFile(filePath: string) {
    return filePath.indexOf('node_modules') !== -1 && filePath.indexOf('@epam/') === -1;
}

export function resolveModuleName(absolutePath: string) {
    const rel = makeRelativeToUuiRoot(path.resolve(absolutePath));
    const firstDir = rel.split('/')[0];
    const pkg = path.resolve(uuiRoot, `${firstDir}/package.json`);
    if (fs.existsSync(pkg)) {
        return JSON.parse(fs.readFileSync(pkg).toString()).name;
    }
    return mapModuleName(rel);
}
function removeTrailingDots(rel: string) {
    /**
     * Converts this: '../../../../node_modules/typescript/lib/lib.es5.d.ts'
     * To this: 'node_modules/typescript/lib/lib.es5.d.ts'
     */
    return rel.replace(/^((..[/])+)(.*)$/g, '$3');
}
function mapModuleName(rel: string): string {
    const map: Record<string, string> = {
        'node_modules/@types/react/index.d.ts': '@types/react',
    };
    const prefix = Object.keys(map).find((key) => rel.startsWith(key));
    if (prefix) {
        const suffix = rel.substring(prefix.length);
        return `${map[prefix]}${suffix}`;
    }
    return rel;
}
