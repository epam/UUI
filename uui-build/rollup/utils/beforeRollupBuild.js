const fs = require('fs-extra');
const { copyPackageJsonAsync } = require('./../../utils/packageJsonUtils');
const path = require('path');

const BUILD_FOLDER = 'build';

module.exports = {
    beforeRollupBuild,
};

/**
 * @param {string[]} copyAsIs
 * @returns {Promise<void>}
 */
async function copyStaticFilesAsync(copyAsIs = []) {
    const p = copyAsIs.map(async c => {
        if (await fs.exists(c)) {
            await fs.copy(c, `${BUILD_FOLDER}/${c}`);
        }
    });
    await Promise.all(p);
}

async function copyPackageJson({ moduleRootDir, packageJsonTransform }) {
    await copyPackageJsonAsync({
        fromDir: moduleRootDir,
        toDir: path.resolve(moduleRootDir, `./${BUILD_FOLDER}`),
        transform: packageJsonTransform,
    });
}

async function beforeRollupBuild({ moduleRootDir, packageJsonTransform, copyAsIs }) {
    fs.emptyDirSync(BUILD_FOLDER);
    await copyStaticFilesAsync(copyAsIs);
    await copyPackageJson({ moduleRootDir, packageJsonTransform });
}
