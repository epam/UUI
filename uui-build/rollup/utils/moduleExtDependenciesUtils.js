const fs = require("fs");
const path = require("path");

module.exports = { getExternalDeps };

const DEPS_ALWAYS_BUNDLED = ["@epam/assets"]

function isImportFromEpamModuleAssets(importId) {
    const epamModuleAssetsRegex = new RegExp('@epam/[^/]+/assets/.+');
    return epamModuleAssetsRegex.test(importId)
}

function getExternalDeps(moduleRootDir) {
    const allKeys = getModuleDepsProd(moduleRootDir)
    const keysExternal = allKeys.filter(key => {
        return !DEPS_ALWAYS_BUNDLED.find(sb => sb === key);
    });
    return importId => {
        return keysExternal.some(keyExternal => {
            /**
             * Sub-folders of "external" package are also external.
             * Exception:
             *      When module imports from "@epam/<module-name>/assets/" folder.
             *      We need to bundle this dependency in such case.
             */
            return keyExternal === importId ||
                (importId.indexOf(`${keyExternal}/`) === 0 && !isImportFromEpamModuleAssets(importId));
        })
    }
}

function getModuleDepsProd(moduleRootDir) {
    const packageJsonPath = path.resolve(moduleRootDir, "package.json");
    const pkg = fs.readFileSync(packageJsonPath).toString();
    const json = JSON.parse(pkg);
    // json.devDependencies is not added here, because it cannot be specified as external one.
    return Object.keys({ ...json.dependencies, ...json.peerDependencies });
}
