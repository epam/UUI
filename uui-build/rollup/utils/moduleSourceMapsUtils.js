const path = require('path');

module.exports = { getSourceMapTransform };

function forwardSlashes(pathStr) {
    return pathStr.replace(/\\/g, '/');
}

/**
 * It's needed to fix sources location path in "build/index.js.map" and "build/styles.css.map".
 */
function getSourceMapTransform({ type, moduleName, moduleFolderName }) {
    return function sourcemapPathTransform(relativeSourcePathParam) {
        const relativeSourcePath = forwardSlashes(relativeSourcePathParam);
        const PROTOCOL = 'rollup://'; // decide what protocol we need to use here "webpack://" or "rollup://" or anything else?
        const BASE_MODULE_PREFIX = `${PROTOCOL}${moduleName}/`;
        const PREFIX = `${BASE_MODULE_PREFIX}./`;
        switch (type) {
            case 'css': {
                /**
                 * It is generated relative to ./build folder of the module.
                 * Known use cases, and how it will be transformed:
                 * a) "sources":["../src/components/widgets/AvatarStack.scss",...] --> "sources":["rollup://<moduleName>/./src/components/widgets/AvatarStack.scss",...]
                 * b) "sources":["../node_modules/@epam/assets/scss/typography.scss",...] --> "sources":["rollup://@epam/assets/scss/promo/typography.scss",...]
                 */
                if (relativeSourcePath.indexOf('../node_modules/') === 0) {
                    return PROTOCOL + relativeSourcePath.substring('../node_modules/'.length);
                }
                return PREFIX + forwardSlashes(path.join('./build', relativeSourcePath));
            }
            case 'js': {
                /**
                 * It's relative to ./build folder, but it's not consistent where it starts. Known cases:
                 * - start from module root ("../src/icons/bold.svg")
                 * - start from monorepo root ("../../src/Test.tsx"). It's because the TS compiler works this way.
                 * Known use cases, and how it will be transformed:
                 * "sources":["../../src/Test.tsx",...] --> "sources":["rollup://<moduleName>/./src/Test.tsx",...]
                 * "sources":["../src/icons/bold.svg",...] --> "sources":["rollup://<moduleName>/./src/icons/bold.svg",...]
                 */

                if (relativeSourcePath.indexOf('../../') === 0) {
                    return PREFIX + forwardSlashes(path.join(`${moduleFolderName}/build`, relativeSourcePath));
                } else if (relativeSourcePath.indexOf('../') === 0) {
                    return PREFIX + forwardSlashes(path.join('./build', relativeSourcePath));
                } else {
                    console.log(relativeSourcePath);
                    return relativeSourcePath;
                }
            }
            default: {
                throw new Error(`Unknown type=${type}`);
            }
        }
    };
}
