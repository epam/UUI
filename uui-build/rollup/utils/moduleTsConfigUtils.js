const fs = require("fs");
const path = require("path");

module.exports = { getTsConfigFile }

const MODULE_TSCONFIG_PATH = './tsconfig.json';

function getTsConfigFile(moduleRootDir) {
    function ensureModuleHasTsConfig(moduleRootDir) {
        const tsconfig = path.resolve(moduleRootDir, MODULE_TSCONFIG_PATH);
        if (!fs.existsSync(tsconfig)) {
            throw new Error(`Unable to find tsconfig: ${tsconfig}.`)
        }
    }
    ensureModuleHasTsConfig(moduleRootDir);
    return path.resolve(moduleRootDir, MODULE_TSCONFIG_PATH)
}
