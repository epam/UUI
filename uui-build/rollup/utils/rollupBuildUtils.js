const { getConfig } = require('../rollup.config')
const rollup = require('rollup');
const { getTsConfigFile } = require("./moduleTsConfigUtils");

module.exports = { buildUsingRollup };

async function buildUsingRollup({ moduleRootDir, moduleIndexFile }) {
    const tsconfigFile = getTsConfigFile(moduleRootDir);
    const cfg = await getConfig({ moduleRootDir, moduleIndexFile, tsconfigFile });
    const { output: outputConfig, ...inputConfig } = cfg[0];
    let bundle;
    const cleanup = async () => {
        bundle && await bundle.close();
    };
    try {
        bundle = await rollup.rollup({ ...inputConfig });
        await Promise.all(outputConfig.map(bundle.write));
    } catch (error) {
        console.error(error);
        process.exit(1);
    } finally {
        await cleanup();
    }
}
