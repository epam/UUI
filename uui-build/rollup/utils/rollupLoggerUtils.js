module.exports = { onwarn }

/**
 * Handles Rollup build warnings.
 * @param message
 */
function onwarn(message) {
    switch (message?.code) {
        case 'CIRCULAR_DEPENDENCY': {
            // skip for now, uncomment to see how many we have.
            // console.warn(message.message)
            break;
        }
        default: {
            console.warn(message.message)
            break;
        }
    }
}
