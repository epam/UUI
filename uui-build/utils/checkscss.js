const { iterateFilesInDirAsync } = require('./fileUtils.js');
const { uuiRoot } = require('./constants.js');
const path = require('path');

async function main() {
    const countNotAssets = [];
    const countAssets = [];

    await iterateFilesInDirAsync(uuiRoot, (fp) => {
        if (fp.indexOf('node_modules') === -1 && fp.indexOf('.git') === -1 && fp.indexOf('build') === -1 && fp.endsWith('.scss')) {
            if (fp.indexOf('assets') !== -1) {
                countAssets.push(path.relative(uuiRoot, fp));
            } else {
                countNotAssets.push(path.relative(uuiRoot, fp));
            }
        }
    }, { recursive: true });

    console.log(`Totals - assets=${countAssets.length} not-assets=${countNotAssets.length}`);

    console.log(`\nAssets: ${countAssets.join(',\n')}\n`);
    console.log(`\nNotAssets: ${countNotAssets.join(',\n')}\n`);
}

main();
