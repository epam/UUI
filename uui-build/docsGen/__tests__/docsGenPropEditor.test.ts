import { generateDocs } from './utils/test-utils';

describe('docsGen:propEditor', () => {
    test('should convert prop editor for optional icon prop', () => {
        const input = `
            type Icon = React.FC<any>;
            export interface TTest {
                icon?: Icon;
            };
        `;
        expect(generateDocs(input)).toMatchSnapshot();
    });
});
