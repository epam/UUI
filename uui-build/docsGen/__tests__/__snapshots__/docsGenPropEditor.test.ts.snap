// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docsGen:propEditor should convert prop editor for optional icon prop 1`] = `
{
 "docsGenTypes": {
  "test/test.tsx:TTest": {
   "summary": {
    "module": "test/test.tsx",
    "typeName": {
     "name": "TTest",
     "nameFull": "TTest"
    },
    "src": "test/test.tsx",
    "exported": true
   },
   "details": {
    "kind": 263,
    "typeValue": {
     "raw": "TTest",
     "print": [
      "interface TTest {",
      "    icon?: Icon;",
      "}"
     ]
    },
    "props": [
     {
      "uid": 1,
      "name": "icon",
      "typeValue": {
       "raw": "any"
      },
      "required": false
     }
    ],
    "propsFromUnion": false
   }
  }
 }
}
`;
